File: ios-native/Sources/CoreNetwork/ChangeEventStream.swift
import CoreDatastore
import Foundation

/// The server's change stream.
///
/// What arrives is an **invalidation signal, not a delta**: `{entity, action,
/// id?}` says what kind of thing changed, never what it became. There is no
/// `changedSince` cursor and no tombstone, so a listener's only correct
/// response is to re-read. That is also why there is no local database — one
/// would not save a single request.
///
/// The Android client gets its SSE parsing from okhttp-sse. `URLSession` has no
/// equivalent, so the framing is done here; it is twenty lines because this
/// server uses one field. The three things that matter are the same:
///
///  1. The stream carries the access token and **renews it**, through the same
///     ``TokenRefresher`` the request path uses. A stream that could not renew
///     would die fifteen minutes in.
///  2. No filtering on the event `type`. This server sends no `event:` field,
///     and its heartbeat is a bare `: ping` comment.
///  3. Reconnection with backoff. The server advertises `retry: 3000` and
///     nothing acts on it unless we do.
public final class ChangeEventStream: Sendable {

    private let serverStore: ServerStore
    private let sessionStore: SessionStore
    private let refresher: TokenRefresher
    private let session: URLSession

    public init(
        serverStore: ServerStore,
        sessionStore: SessionStore,
        refresher: TokenRefresher
    ) {
        self.serverStore = serverStore
        self.sessionStore = sessionStore
        self.refresher = refresher

        let configuration = URLSessionConfiguration.default
        // The stream is meant to stay open; the default 60 s request timeout
        // would kill it between two changes. This is the counterpart of the
        // Android client's `readTimeout(0)`.
        configuration.timeoutIntervalForRequest = .greatestFiniteMagnitude
        configuration.timeoutIntervalForResource = .greatestFiniteMagnitude
        configuration.waitsForConnectivity = true
        self.session = URLSession(configuration: configuration)
    }

    /// Reconnects on its own until the consumer stops iterating.
    public func changes() -> AsyncStream<TodoChangeEvent> {
        AsyncStream { continuation in
            let task = Task {
                var attempt = 0
                while !Task.isCancelled {
                    do {
                        let outcome = try await self.consume(into: continuation)
                        switch outcome {
                        case .closedByServer:
                            // A clean close is not a failure and is not retried
                            // — same call `retryWhen` makes on the Android side
                            // when okhttp reports `onClosed`.
                            continuation.finish()
                            return
                        case .unauthorized:
                            // Renew and reconnect straight away: this is an
                            // expired token, not an unhealthy server, and there
                            // is nothing to back off from.
                            let stale = await self.sessionStore.accessToken
                            guard await self.refresher.freshToken(
                                replacing: stale
                            ) != nil else {
                                continuation.finish()
                                return
                            }
                            attempt = 0
                            continue
                        }
                    } catch is CancellationError {
                        break
                    } catch {
                        // Anything else: the connection dropped or the server
                        // is unwell. Back off and try again.
                    }

                    try? await Task.sleep(for: .milliseconds(Self.backoffMilliseconds(attempt)))
                    attempt += 1
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    // MARK: - Internals

    private enum Outcome {
        case closedByServer
        case unauthorized
    }

    private func consume(
        into continuation: AsyncStream<TodoChangeEvent>.Continuation
    ) async throws -> Outcome {
        let base = await serverStore.baseURL
        guard let url = APIURL.resolve(base: base, path: "api/v1/events") else {
            throw APIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        if let token = await sessionStore.accessToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let (bytes, response) = try await session.bytes(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw APIError.transport(message: "Not an HTTP response.")
        }
        if http.statusCode == 401 { return .unauthorized }
        guard (200 ..< 300).contains(http.statusCode) else {
            throw APIError.http(status: http.statusCode, body: Data())
        }

        let decoder = JSONDecoder()
        var frames = SSEFrameDecoder()

        for try await line in bytes.lines {
            guard let payload = frames.feed(line) else { continue }
            guard let event = try? decoder.decode(
                TodoChangeEvent.self,
                from: Data(payload.utf8)
            ) else {
                // A frame this build cannot parse is a contract drift, not a
                // reason to tear down the stream.
                continue
            }
            continuation.yield(event)
        }

        return .closedByServer
    }

    /// 1 s, 2 s, 4 s … capped at 30 — the same ramp as the Android client.
    private static func backoffMilliseconds(_ attempt: Int) -> Int {
        min(1_000 << min(attempt, 5), 30_000)
    }
}

File: android-native/core/network/src/main/kotlin/today/mindlog/todo/core/network/ChangeEventStream.kt
package today.mindlog.todo.core.network

import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.retryWhen
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.sse.EventSource
import okhttp3.sse.EventSourceListener
import okhttp3.sse.EventSources
import today.mindlog.todo.core.network.di.ApiClient
import today.mindlog.todo.core.network.model.ChangeEvent
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.min

/**
 * The server's change stream.
 *
 * What arrives is an **invalidation signal, not a delta**: `{entity, action,
 * id?}` says what kind of thing changed, never what it became. There is no
 * `changedSince` cursor and no tombstone, so a listener's only correct
 * response is to re-read. That is also why there is no local database yet — one
 * would not save a single request.
 *
 * Adapted from the archived talk client's `EventStream.kt`, with three changes
 * that matter:
 *
 *  1. The client is **derived** from the authenticated one rather than built
 *     fresh, so the stream inherits the auth interceptor *and* the 401 renewal.
 *     A stream that could not renew would die fifteen minutes in.
 *  2. No filtering on the event `type`. This server sends no `event:` field, so
 *     the type is always null; its heartbeat is a bare `: ping` comment, which
 *     okhttp-sse already discards.
 *  3. Reconnection with backoff. The server advertises `retry: 3000`, but
 *     okhttp-sse does not act on it — nothing reconnects unless we do.
 */
@Singleton
class ChangeEventStream @Inject constructor(
    @ApiClient okHttpClient: OkHttpClient,
    private val json: Json,
) {
    private val sseClient: OkHttpClient = okHttpClient.newBuilder()
        // The stream is meant to stay open; any read timeout would kill it
        // between two changes.
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()

    fun changes(): Flow<ChangeEvent> = rawEvents().retryWhen { cause, attempt ->
        if (cause is CancellationSignal) return@retryWhen false
        delay(min(1_000L shl attempt.toInt().coerceAtMost(5), MAX_BACKOFF_MS))
        true
    }

    private class CancellationSignal : Exception()

    private fun rawEvents(): Flow<ChangeEvent> = callbackFlow {
        val request = Request.Builder()
            .url("${BaseUrlInterceptor.PLACEHOLDER_BASE_URL}api/v1/events")
            .header("Accept", "text/event-stream")
            .build()

        val listener = object : EventSourceListener() {
            override fun onEvent(
                eventSource: EventSource,
                id: String?,
                type: String?,
                data: String,
            ) {
                runCatching { json.decodeFromString<ChangeEvent>(data) }
                    .onSuccess { trySend(it) }
                // A frame we cannot parse is a contract drift, not a reason to
                // tear down the stream.
            }

            override fun onClosed(eventSource: EventSource) {
                close(CancellationSignal())
            }

            override fun onFailure(eventSource: EventSource, t: Throwable?, response: Response?) {
                close(t ?: IllegalStateException("SSE stream failed: ${response?.code}"))
            }
        }

        val source = EventSources.createFactory(sseClient).newEventSource(request, listener)
        awaitClose { source.cancel() }
    }

    private companion object {
        const val MAX_BACKOFF_MS = 30_000L
    }
}
